mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxViewModelProvider.kt

```diff
@@ -22,68 +26,81 @@ object MvRxViewModelProvider {
         viewModelContext: ViewModelContext,
         key: String = viewModelClass.name,
+        forExistingViewModel: Boolean = false,
         initialStateFactory: MvRxStateFactory<VM, S> = RealMvRxStateFactory()
     ): VM {
-        // This wraps the fact that ViewModelProvider.of has individual methods for Fragment and FragmentActivity.
-        val factory = MvRxFactory { createViewModel(viewModelClass, stateClass, viewModelContext, initialStateFactory = initialStateFactory) }
-        return when (viewModelContext) {
-            is ActivityViewModelContext -> ViewModelProviders.of(viewModelContext.activity, factory)
-            is FragmentViewModelContext -> ViewModelProviders.of(viewModelContext.fragment, factory)
-        }.get(key, viewModelClass)
-    }
+        val stateRestorer = viewModelContext
+            .savedStateRegistry
+            .consumeRestoredStateForKey(key)
+            ?.toStateRestorer<S>(viewModelContext)
 
-    @Suppress("UNCHECKED_CAST")
-    internal fun <VM : BaseMvRxViewModel<S>, S : MvRxState> createViewModel(
-        viewModelClass: Class<out VM>,
-        stateClass: Class<out S>,
-        viewModelContext: ViewModelContext,
-        stateRestorer: (S) -> S = { it },
-        initialStateFactory: MvRxStateFactory<VM, S> = RealMvRxStateFactory()
-    ): VM {
-        val initialState = initialStateFactory.createInitialState(viewModelClass, stateClass, viewModelContext, stateRestorer)
-        val factoryViewModel = viewModelClass.factoryCompanion()?.let { factoryClass ->
-            try {
-                factoryClass.getMethod("create", ViewModelContext::class.java, MvRxState::class.java)
-                    .invoke(factoryClass.instance(), viewModelContext, initialState) as VM?
-            } catch (exception: NoSuchMethodException) {
-                // Check for JvmStatic method.
-                viewModelClass.getMethod("create", ViewModelContext::class.java, MvRxState::class.java)
-                    .invoke(null, viewModelContext, initialState) as VM?
+        val restoredContext = stateRestorer?.viewModelContext ?: viewModelContext
+
+        val viewModel = ViewModelProvider(
+            viewModelContext.owner,
+            MvRxFactory(
+                viewModelClass,
+                stateClass,
+                restoredContext,
+                key,
+                stateRestorer?.toRestoredState,
+                forExistingViewModel,
+                initialStateFactory
+            )
+        ).get(key, viewModelClass)
+
+        try {
+            // Save the view model's state to the bundle so that it can be used to recreate
+            // state across system initiated process death.
+            viewModelContext.savedStateRegistry.registerSavedStateProvider(key) {
+                viewModel.getSavedStateBundle(restoredContext.args)
             }
+        } catch (e: IllegalArgumentException) {
+            // The view model was already registered with the context.
```

Comment:
can you add some more details here about why this is expected and ok? and about how we need to ensure that only the initial fragment that created the viewmodel is used to save state?